Docker container requires writable /deploy directory for metrics

An issue was found  running the stock Tiled container on the Spin infrastructure. The /deploy directory was mounted as read only, and the user was something other than root. Tiled fails to start.

I can reproduce locally using the instructions at https://blueskyproject.io/tiled/how-to/docker.html but changing the documentation to use ghcr image (which should probably be in the documentation).

```docker run --rm -p 8000:8000   --mount type=bind,source="$(pwd)/example_files",target=/deploy/example_files  --mount type=bind,source="$(pwd)/",target=/deploy   --env TILED_CONFIG=/deploy/config.yml ghcr.io/bluesky/tiled:main```


Tiled fails to start, because the it's requiring a writable /deploy directory for prometheus. The fix for spin was to make the `/deploy` mount writable, but that doesn't feel right. Not everyone will use prometheus and it is reasonable to make this directory read only. It feels like if you want to use prometheus, than sure, make it writeable. 

So what's the right fix? Make prometheus an option that you can turn on if you want?

```
[2021-12-16 23:30:38 +0000] [20] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/opt/venv/lib/python3.8/site-packages/uvicorn/workers.py", line 66, in init_process
    super(UvicornWorker, self).init_process()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/util.py", line 359, in import_app
    mod = importlib.import_module(module)
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/app/tiled/server/app.py", line 33, in <module>
    from .metrics import capture_request_metrics
  File "/app/tiled/server/metrics.py", line 77, in <module>
    REQUEST_DURATION.labels(code=code, method="GET", endpoint=endpoint)
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/metrics.py", line 183, in labels
    self._metrics[labelvalues] = self.__class__(
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/metrics.py", line 548, in __init__
    super(Histogram, self).__init__(
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/metrics.py", line 131, in __init__
    self._metric_init()
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/metrics.py", line 576, in _metric_init
    self._sum = values.ValueClass(self._type, self._name, self._name + '_sum', self._labelnames, self._labelvalues)
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/values.py", line 70, in __init__
    self.__reset()
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/values.py", line 84, in __reset
    files[file_prefix] = MmapedDict(filename)
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/mmap_dict.py", line 61, in __init__
    self._f = open(filename, 'rb' if read_mode else 'a+b')
FileNotFoundError: [Errno 2] No such file or directory: '/deploy/prometheus/histogram_20.db'
[2021-12-16 23:30:38 +0000] [20] [INFO] Worker exiting (pid: 20)
[2021-12-16 23:30:38 +0000] [1] [WARNING] Worker with pid 20 was terminated due to signal 15
[2021-12-16 23:30:38 +0000] [1] [WARNING] Worker with pid 31 was terminated due to signal 15
Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 209, in run
    self.sleep()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 357, in sleep
    ready = select.select([self.PIPE[0]], [], [], 1.0)
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/venv/bin/gunicorn", line 8, in <module>
    sys.exit(run())
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 67, in run
    WSGIApplication("%(prog)s [OPTIONS] [APP_MODULE]").run()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/app/base.py", line 231, in run
    super().run()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/app/base.py", line 72, in run
    Arbiter(self).run()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 229, in run
    self.halt(reason=inst.reason, exit_status=inst.exit_status)
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 342, in halt
    self.stop()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 393, in stop
    time.sleep(0.1)
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 540, in reap_workers
    self.cfg.child_exit(self, worker)
  File "/app/docker/gunicorn_config.py", line 6, in child_exit
    prometheus_client.multiprocess.mark_process_dead(worker.pid)
  File "/opt/venv/lib/python3.8/site-packages/prometheus_client/multiprocess.py", line 161, in mark_process_dead
    for f in glob.glob(os.path.join(path, 'gauge_livesum_{0}.db'.format(pid))):
  File "/usr/local/lib/python3.8/glob.py", line 21, in glob
    return list(iglob(pathname, recursive=recursive))
  File "/usr/local/lib/python3.8/glob.py", line 34, in iglob
    sys.audit("glob.glob", pathname, recursive)
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/opt/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
```

--- a/docker/Dockerfile
+++ b/docker/Dockerfile
@@ -28,9 +28,6 @@ COPY --from=builder /app /app
 
 WORKDIR /deploy
 
-RUN mkdir /deploy/prometheus
-ENV PROMETHEUS_MULTIPROC_DIR=/deploy/prometheus
-
 EXPOSE 8000
 
 ENTRYPOINT ["/app/docker/entrypoint.sh"]

--- a/tiled/adapters/files.py
+++ b/tiled/adapters/files.py
@@ -106,6 +106,8 @@ class DirectoryAdapter(MapAdapter):
         # These are unofficial but common file extensions.
         ".hdf": "application/x-hdf5",
         ".hdf5": "application/x-hdf5",
+        # on opensuse csv -> text/x-comma-separated-values
+        ".csv": "text/csv",
     }
 
     @classmethod

--- a/tiled/config.py
+++ b/tiled/config.py
@@ -154,6 +154,27 @@ def construct_serve_tree_kwargs(
         server_settings = {}
         server_settings["allow_origins"] = config.get("allow_origins")
         server_settings["object_cache"] = config.get("object_cache", {})
+
+        metrics = config.get("metrics", {})
+        if metrics.get("prometheus", False):
+            prometheus_multiproc_dir = os.getenv("PROMETHEUS_MULTIPROC_DIR", None)
+            if not prometheus_multiproc_dir:
+                raise ValueError(
+                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR env variable not set"
+                )
+            elif not Path(prometheus_multiproc_dir).is_dir():
+                raise ValueError(
+                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR"
+                    f"({prometheus_multiproc_dir}) is not a directory"
+                )
+            elif not os.access(prometheus_multiproc_dir, os.W_OK):
+                raise ValueError(
+                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR"
+                    f"({prometheus_multiproc_dir}) is not writable"
+                )
+
+        server_settings["metrics"] = metrics
+
         for structure_family, values in config.get("media_types", {}).items():
             for media_type, import_path in values.items():
                 serializer = import_object(import_path, accept_live_object=True)
@@ -182,6 +203,7 @@ def merge(configs):
     access_control_config_source = None
     uvicorn_config_source = None
     object_cache_config_source = None
+    metrics_config_source = None
     allow_origins = []
     media_types = defaultdict(dict)
     file_extensions = {}
@@ -228,6 +250,15 @@ def merge(configs):
                 )
             object_cache_config_source = filepath
             merged["object_cache"] = config["object_cache"]
+        if "metrics" in config:
+            if "metrics" in merged:
+                raise ConfigError(
+                    "metrics can only be specified in one file. "
+                    f"It was found in both {metrics_config_source} and "
+                    f"{filepath}"
+                )
+            metrics_config_source = filepath
+            merged["metrics"] = config["metrics"]
         for item in config.get("trees", []):
             if item["path"] in paths:
                 msg = "A given path may be only be specified once."

--- a/tiled/config.py
+++ b/tiled/config.py
@@ -164,12 +164,12 @@ def construct_serve_tree_kwargs(
                 )
             elif not Path(prometheus_multiproc_dir).is_dir():
                 raise ValueError(
-                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR"
+                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR "
                     f"({prometheus_multiproc_dir}) is not a directory"
                 )
             elif not os.access(prometheus_multiproc_dir, os.W_OK):
                 raise ValueError(
-                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR"
+                    "prometheus enabled but PROMETHEUS_MULTIPROC_DIR "
                     f"({prometheus_multiproc_dir}) is not writable"
                 )
 

--- a/tiled/schemas/service_configuration.yml
+++ b/tiled/schemas/service_configuration.yml
@@ -384,4 +384,6 @@ properties:
       prometheus:
         type: boolean
         description: |
-          Enable/Disable prometheus metrics
+          Enable/Disable prometheus metrics. Default is false.
+          If enabled `PROMETHEUS_MULTIPROC_DIR` environment variable
+          must be set to the path of a writable directory.

--- a/tiled/schemas/service_configuration.yml
+++ b/tiled/schemas/service_configuration.yml
@@ -378,3 +378,10 @@ properties:
       ssl_ciphers:
         type: string
         description: Ciphers to use (see stdlib ssl module's). Default TLSv1.
+  metrics:
+    type: object
+    properties:
+      prometheus:
+        type: boolean
+        description: |
+          Enable/Disable prometheus metrics

--- a/tiled/server/app.py
+++ b/tiled/server/app.py
@@ -30,7 +30,6 @@ from .core import (
     get_serialization_registry,
     record_timing,
 )
-from .metrics import capture_request_metrics
 from .object_cache import NO_CACHE, ObjectCache
 from .object_cache import logger as object_cache_logger
 from .object_cache import set_object_cache
@@ -155,10 +154,7 @@ def get_app(
     @app.middleware("http")
     async def capture_metrics(request: Request, call_next):
         """
-        This does two things:
-
-        - Place metrics in Server-Timing header, in accordance with HTTP spec.
-        - Capture metrics in prometheus. The proemetheus metrics are availabe from /metrics.
+        Place metrics in Server-Timing header, in accordance with HTTP spec.
         """
         # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing
         # https://w3c.github.io/server-timing/#the-server-timing-header-field
@@ -189,7 +185,6 @@ def get_app(
             for key, metrics_ in metrics.items()
         )
         response.__class__ = PatchedStreamingResponse  # tolerate memoryview
-        capture_request_metrics(request, response)
         return response
 
     @app.middleware("http")
@@ -337,6 +332,7 @@ def serve_tree(
     include_routers = []
     include_routers.extend(getattr(tree, "include_routers", []))
     include_routers.extend(getattr(authenticator, "include_routers", []))
+
     # Likewise, the Tree and Authenticator can run tasks in the background.
     # These typically contain a periodic loop.
     background_tasks = []
@@ -367,6 +363,19 @@ def serve_tree(
         app.dependency_overrides[
             get_serialization_registry
         ] = override_get_serialization_registry
+
+    metrics_config = server_settings.get("metrics", {})
+    if metrics_config.get("prometheus", False):
+        from . import metrics
+
+        app.include_router(metrics.router)
+
+        @app.middleware("http")
+        async def capture_metrics_prometheus(request: Request, call_next):
+            response = await call_next(request)
+            metrics.capture_request_metrics(request, response)
+            return response
+
     return app
 
 

--- a/tiled/server/metrics.py
+++ b/tiled/server/metrics.py
@@ -4,7 +4,12 @@ conventions for metrics & labels. We generally prefer naming them
 `tiled_<noun>_<verb>_<type_suffix>`.
 """
 
+from functools import lru_cache
+
+from fastapi import APIRouter, Request, Response
 from prometheus_client import Histogram
+
+router = APIRouter()
 
 REQUEST_DURATION = Histogram(
     "tiled_request_duration_seconds",
@@ -124,3 +129,32 @@ def capture_request_metrics(request, response):
         COMPRESSION_RATIO.labels(
             method=method, code=code, endpoint=endpoint, encoding=encoding
         ).observe(metrics["compress"]["ratio"])
+
+
+@lru_cache()
+def prometheus_registry():
+    """
+    Configure prometheus_client.
+
+    This is run the first time the /metrics endpoint is used.
+    """
+    # The multiprocess configuration makes it compatible with gunicorn.
+    # https://github.com/prometheus/client_python/#multiprocess-mode-eg-gunicorn
+    from prometheus_client import CollectorRegistry
+    from prometheus_client.multiprocess import MultiProcessCollector
+
+    registry = CollectorRegistry()
+    MultiProcessCollector(registry)  # This has a side effect, apparently.
+    return registry
+
+
+@router.get("/metrics")
+async def metrics(request: Request):
+    """
+    Prometheus metrics
+    """
+    from prometheus_client import CONTENT_TYPE_LATEST, generate_latest
+
+    request.state.endpoint = "metrics"
+    data = generate_latest(prometheus_registry())
+    return Response(data, headers={"Content-Type": CONTENT_TYPE_LATEST})

--- a/tiled/server/router.py
+++ b/tiled/server/router.py
@@ -1,10 +1,9 @@
 import dataclasses
 import inspect
 from datetime import datetime, timedelta
-from functools import lru_cache
 from typing import Any, List, Optional
 
-from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response
+from fastapi import APIRouter, Depends, HTTPException, Query, Request
 from jmespath.exceptions import JMESPathError
 from pydantic import BaseSettings
 
@@ -101,35 +100,6 @@ async def about(
         resolve_media_type(request),
         expires=datetime.utcnow() + timedelta(seconds=600),
     )
-
-
-@lru_cache()
-def prometheus_registry():
-    """
-    Configure prometheus_client.
-
-    This is run the first time the /metrics endpoint is used.
-    """
-    # The multiprocess configuration makes it compatible with gunicorn.
-    # https://github.com/prometheus/client_python/#multiprocess-mode-eg-gunicorn
-    from prometheus_client import CollectorRegistry
-    from prometheus_client.multiprocess import MultiProcessCollector
-
-    registry = CollectorRegistry()
-    MultiProcessCollector(registry)  # This has a side effect, apparently.
-    return registry
-
-
-@router.get("/metrics")
-async def metrics(request: Request):
-    """
-    Prometheus metrics
-    """
-    from prometheus_client import CONTENT_TYPE_LATEST, generate_latest
-
-    request.state.endpoint = "metrics"
-    data = generate_latest(prometheus_registry())
-    return Response(data, headers={"Content-Type": CONTENT_TYPE_LATEST})
 
 
 def declare_search_router(query_registry):

